include/ignition/math/detail/Capsule.hh

```diff
@@ -0,0 +1,157 @@
+{
+  this->radius = _radius;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+T Capsule<T>::Length() const
+{
+  return this->length;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+void Capsule<T>::SetLength(const T _length)
+{
+  this->length = _length;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+const Material &Capsule<T>::Mat() const
+{
+  return this->material;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+void Capsule<T>::SetMat(const Material &_mat)
+{
+  this->material = _mat;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+bool Capsule<T>::operator==(const Capsule &_capsule) const
+{
+  return equal(this->radius, _capsule.Radius()) &&
+    equal(this->length, _capsule.Length()) &&
+    this->material == _capsule.Mat();
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+bool Capsule<T>::MassMatrix(MassMatrix3<T> &_massMat) const
+{
+  // mass and moment of inertia of cylinder about centroid
+  MassMatrix3<T> cylinder;
+  cylinder.SetFromCylinderZ(this->material, this->length, this->radius);
+
+  // mass and moment of inertia of hemisphere about centroid
+  const T r2 = this->radius * this->radius;
+  const T hemisphereMass = this->material.Density() *
+      2. / 3. * IGN_PI * r2 * this->radius;
+  // efunda.com/math/solids/solids_display.cfm?SolidName=Hemisphere
+  const T ixx = 83. / 320. * hemisphereMass * r2;
+  const T izz = 2. / 5.  * hemisphereMass * r2;
+  MassMatrix3<T> hemisphere(hemisphereMass, Vector3<T>(ixx, ixx, izz),
+                            Vector3<T>::Zero);;
+
+  // Distance from centroid of cylinder to centroid of hemisphere,
+  // since centroid of hemisphere is 3/8 radius from its flat base
+  const T dz = this->length / 2. + this->radius * 3. / 8.;
+  Inertial<T> upperCap(hemisphere, Pose3<T>(0, 0, dz, 0, 0, 0));
+  Inertial<T> lowerCap(hemisphere, Pose3<T>(0, 0, -dz, 0, 0, 0));
+
+  // Use Inertial class to add MassMatrix3 objects at different poses
+  Inertial<T> capsule =
+      Inertial<T>(cylinder, Pose3<T>::Zero) + upperCap + lowerCap;
+
+  _massMat = capsule.MassMatrix();
+  return _massMat.IsValid();
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+T Capsule<T>::Volume() const
+{
+  return IGN_PI * std::pow(this->radius, 2) *
+         (this->length + 4. / 3. * this->radius);
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+bool Capsule<T>::SetDensityFromMass(const T _mass)
+{
+  T newDensity = this->DensityFromMass(_mass);
+  if (newDensity > 0)
+    this->material.SetDensity(newDensity);
+  return newDensity > 0;
+}
+
+//////////////////////////////////////////////////
+template<typename T>
+T Capsule<T>::DensityFromMass(const T _mass) const
+{
+  if (this->radius <= 0 || this->length <=0 || _mass <= 0)
+    return -1.0;
```

Comment:
I know this is following the pattern from other shapes, but I think we should consider returning `NaN` in such cases instead of `-1`.

Reply:
what do you think @azeey?

Reply:
I'm not confident if returning a `NaN` is good practice, but the math functions in the standard c++ library either throw an exception or return a `NaN` when the input is invalid, so I think it's reasonable to return a `NaN`.

Reply:
we don't use `NaN`s much in our APIs, and I'm careful about introducing them because they can quickly spread and it can be hard to figure out where they came from. In this case, I think it's fine because this function isn't widely used

Reply:
2628fabf38785957845ac985d29ea0b3cce07613